### ISSUE TITLE
kotlinx.coroutines.flow.internal.ChildCancelledException: Child of the scoped flow was cancelled

### ISSUE THREAD

**Issue by Thomas-Vos** [Kotlin/kotlinx.coroutines#1433]
Hi, in my app I randomly get a `ChildCancellationException`. It comes from the `flatMapLatest` operator (previously named `switchMap`). I tested it on the JVM with versions 1.3.0-M2, 1.3.0-RC, 1.3.0-RC2. I was able to reproduce it in a test:
```kotlin
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.channels.ConflatedBroadcastChannel
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlin.test.Test

class MyTest {

    @Test
    fun test(): Unit = runBlocking {
        val context = Dispatchers.Default

        val deferred = GlobalScope.async(context) {
            launch(context) {
                observeFlow().collect {}
            }

            while (true) {
                channel.send(channel.value)
            }
        }

        deferred.await()
    }

    val channel = ConflatedBroadcastChannel("")
    val flow = channel.asFlow()

    fun observeFlow() = flow.flatMapLatest {
        val flows = List(2) {
            flow {
                while (true) {
                    emit("")
                }
            }
        }
        combine(flows) { it.asList() }
    }
}
```
```
Exception in thread "DefaultDispatcher-worker-5 @coroutine#716" kotlinx.coroutines.flow.internal.ChildCancelledException: Child of the scoped flow was cancelled
	(Coroutine boundary)
	at kotlinx.coroutines.channels.AbstractChannel.registerSelectReceiveOrNull(AbstractChannel.kt:753)
	at kotlinx.coroutines.channels.AbstractChannel.access$registerSelectReceiveOrNull(AbstractChannel.kt:484)
	at kotlinx.coroutines.channels.AbstractChannel$onReceiveOrNull$1.registerSelectClause1(AbstractChannel.kt:732)
	at kotlinx.coroutines.selects.SelectBuilderImpl.invoke(Select.kt:415)
	at kotlinx.coroutines.flow.internal.CombineKt$combineInternal$2.invokeSuspend(Combine.kt:151)
Caused by: kotlinx.coroutines.flow.internal.ChildCancelledException: Child of the scoped flow was cancelled
	at kotlinx.coroutines.flow.internal.ChannelFlowTransformLatest$flowCollect$3$invokeSuspend$$inlined$collect$1.emit(Collect.kt:137)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt.emitAll(Channels.kt:56)
	at kotlinx.coroutines.flow.FlowKt.emitAll(Unknown Source)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt$emitAll$1.invokeSuspend(Channels.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:241)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:740)
```
Of course this test is not my real code, but this appears to reproduce the exception (almost) every time it is run. You may need to run the test again if you do not get the exception. It should happen within a few seconds of running the test.

Is this expected behaviour and/or is there anything I can do to fix this?

There was a discussion about this on Slack with @qwwdfsad: https://kotlinlang.slack.com/archives/C1CFAFJSK/p1563388954310700

**Comment by qwwdfsad** [comment]
Apparently, it is a bug in the underlying implementation mechanism of `combine`
Will be fixed in 1.3.0, thanks for the repro

**Comment by qwwdfsad** [comment]
It's better to open a new issue with a reproducer I can run

**Comment by psteiger** [comment]
I still see this on 1.3.0. I'm not sure if I'm doing something wrong, but the use case is very similar to the one described.

**Comment by qwwdfsad** [comment]
`offer` is for non-suspending context, while `send` is for suspending ones.
`offer` is, unfortunately, non-symmetric to `send` in terms of propagated exceptions (`CancellationException` from `send` is usually ignored, while `CancellationException` from offer in nom-suspending context is not).
We hope to fix it in #974 either with `offerOrClosed` or changing `offer` semantics

**Comment by AChep** [comment]
I still have the same exception when combining the flows:
```
accessManagerList.all
        .map { it.validatorFlow }
        .fold(flowOf<AccessValidator>(AccessValidatorNone())) { y, x ->
            y.combine(x) { yValidator, xValidator ->
                // Create a new validator that combines two.
                if (xValidator is AccessValidatorAll)
                    return@combine xValidator
                if (yValidator is AccessValidatorAll)
                    return@combine yValidator
                { accessGroup: AccessGroup ->
                    yValidator(accessGroup) || xValidator(accessGroup)
                }
            }
        }
        .flowOn(Dispatchers.Default)
```
after this, the flow is shared and used in A LOT of other flows. Unfortunately, I wasn't able to create a synthetic reproducible case...  all I know is that the crash stops occurring once I change the dispatcher to `Dispatchers.Main`
